src/utils/decorators/tooltip-decorator/tooltip-decorator.js

```diff
@@ -163,198 +122,145 @@ const TooltipDecorator = (ComposedComponent) => {
       }
     }
 
-    /**
-     * A lifecycle called immediatly before new props cause a re-render
-     * Resets the hover state if active
-     *
-     * @method componentWillReceiveProps
-     */
     componentWillReceiveProps(nextProps) {
       if (super.componentWillReceiveProps) { super.componentWillReceiveProps(nextProps); }
 
-      if (this.isVisible()) {
-        this.setState({ isVisible: false });
+      if (nextProps.tooltipPosition !== this.props.tooltipPosition) {
+        this.setState({ tooltipPosition: '', tooltipAlign: '' });
       }
     }
 
     state = {
-      /**
-       * Whether tooltip currently showing
-       *
-       * @property isVisible
-       * @type {Boolean}
-       * @default false
-       */
-      isVisible: false
+      isVisible: false,
+      tooltipAlign: '',
+      tooltipPosition: ''
     };
 
     isVisible = () => {
       return this.state.isVisible || this.props.tooltipVisible;
     }
 
-    /**
-     * Shows tooltip
-     *
-     * @method onShow
-     * @return {void}
-     */
     onShow = () => {
+      clearTimeout(this._showTooltipTimeout);
       clearTimeout(this._hideTooltipTimeout);
 
+      if (this.state.isVisible) return;
+
       this._showTooltipTimeout = setTimeout(() => {
         this.setState({ isVisible: true });
         this.positionTooltip();
       }, 100);
     };
 
-    /**
-     * Hides tooltip
-     *
-     * @method onHide
-     * @return {void}
-     */
     onHide = () => {
       clearTimeout(this._showTooltipTimeout);
+      clearTimeout(this._hideTooltipTimeout);
+
+      if (!this.state.isVisible) return;
 
       this._hideTooltipTimeout = setTimeout(() => {
         this.setState({ isVisible: false });
       }, 100);
     };
 
-    /**
-     * Returns the DOM node of the target.
-     *
-     * @method getTarget
-     * @return {DOM node}
-     */
     getTarget = () => {
       return ReactDOM.findDOMNode(this._target); // eslint-disable-line react/no-find-dom-node
     }
 
-    /**
-     * Returns the DOM node of the tooltip.
-     *
-     * @method getTooltip
-     * @return {DOM node}
-     */
     getTooltip = () => {
       return ReactDOM.findDOMNode(this._tooltip); // eslint-disable-line react/no-find-dom-node
     }
 
-    /**
-     * Calculates position for tooltip, first result cached.
-     *
-     * @method calculatePosition
-     * @param {Object} tooltip
-     * @param {Object} target
-     * @return {Object} shifts calculated
-     */
     calculatePosition = (tooltip, target) => {
```

Comment:
To me, these functions – `calculatePosition()` and `positionTooltip()` – seem unnecessarily complicated.

In my [`FE-1847-tooltip-positioning`](https://github.com/Sage/carbon/commits/FE-1847-tooltip-positioning) branch, I simplified / cleaned up those functions and I moved them to a new file – see [`positionTooltip()` in `tooltip-position.config.js`](https://github.com/Sage/carbon/blob/1a0afac/src/components/tooltip/tooltip-position.config.js#L28-L80).

I think it'd be good to simplify / clean up the code in this branch in the same way.

Reply:
I have moved `calculatePosition` into a separate file and simplified `positionTooltip` method, unfortunately I had to add more logic, based on new requirements, so these methods could not be simplified in the same way.

Reply:
OK. Thanks for moving `calculatePosition()` to the new file `src/utils/decorators/tooltip-decorator/calculate-position.js` :white_check_mark: